lib/tests/triples.nix

```diff
@@ -0,0 +1,157 @@
+# rare and unusual cases where nixpkgs has carefully and
+# deliberately decided to diverge from it.
+#
+# We use `gnu-config` as our source of ground truth not because it
+# is elegant (it isn't) or sensible (it isn't), but because it:
+#
+# 1. has an executable (i.e. unambiguous) specification
+# 2. covers a superset of the systems on which nixpkgs could be useful
+# 3. is *not* under nixpkgs' control (bickering/drama avoidance)
+# 4. emanates from an organization (GNU) which understands that
+#    the taxonomy is not merely an implementation detail of one of
+#    their software products, and has demonstrated this
+#    understanding over an extended period of time (multiple
+#    decades).
+#
+
+let
+  lib  = import ../default.nix;
+  pkgs = import ../.. {};
+
+  inherit (lib.systems.parse) canonicalize;
+
+  filtered-triples = lib.pipe null [
+
+    # start with the lists of known possibilities for each field
+    (_: with lib.systems.parse; {
+      cpu    = cpuTypes;
+      vendor = vendors;
+      kernel = kernels;
+      abi    = abis;
+    })
+
+    # drop the structure; keep only the `name` attribute for each possibility
+    (builtins.mapAttrs (_: lib.mapAttrsToList (k: v: v.name or k)))
+
+    # generate every possible {cpu,vendor,kernel,abi} combination
+    lib.cartesianProductOfSets
+
+    # Then form a string triple for each of them
+    (map lib.systems.parse.tripleFromSkeleton)
+
+    # Filter out any triples for which nixpkgs throws while
+    # attempting to parse.  This makes the tests much faster since
+    # gnu-config needs to fork() a shell for each test case.  This is
+    # safe since it is okay for nixpkgs' set of acceptable triples to
+    # be a subset of gnu-config's set of acceptable triples.
+    (lib.filter (triple:
+      (builtins.tryEval (canonicalize triple))
+        .success))
+
+    #
+    # The following are triples which we have deliberately
+    # chosen to accept differently than gnu-config.  Entries here
+    # impose an ongoing burden on nixpkgs, so PRs which add new
+    # triples here should allow for a longer review period and
+    # should be advertised widely.
+    #
+    (lib.filter (v:
+
+      # gnu-config infers the vendor "pc" for these Nix doubles, but
+      # nixpkgs has used ${cpu}-unknown-linux-gnu for a very long time.
+      v != "x86_64-linux" && v != "i686-linux" &&
+      v != "x86_64-linux-gnu" && v != "i686-linux-gnu" &&
+      v != "x86_64-darwin" && v != "aarch64-darwin" &&
+
+      # We have been shipping Broadcom VC4 forks of gcc/binutils
+      # using nonstandard triples since 4aa1ffae041bb9c65eb3067e9dbaaa70710ed100
+      (canonicalize v) != "vc4-elf"
+
+      # This triple is special to GHC/Cabal/GHCJS and not recognized by autotools
+      # See: https://gitlab.haskell.org/ghc/ghc/-/commit/6636b670233522f01d002c9b97827d00289dbf5c
+      # https://github.com/ghcjs/ghcjs/issues/53
+      && (canonicalize v) != "javascript-unknown-ghcjs"
```

Comment:
I probably should have titled this PR "add test suite **and fix our broke-ass parser so it passes the tests**".

This was a lot of work.  Very boring and tedious work.

> The general principle for me is that it is fine for either us or GNU config to reject something the other accepts, but they shouldn't both accept something and normalize it in a different way.

That is what this PR does, with five clearly enumerated exceptions (see below).

> For the `system` field GNU config and us already agree, right?

Without this PR?  No, not even close.

The only exceptions are listed above, and this PR puts the all in one place so it's easy to know "where do we behave differently than `gnu-config`"?  Fixing the first three would be a very jarring adjustment (and massive rebuild); I'm not opposed to that but if it is to be done it should happen in a separate PR after this one.  The last exception was a deliberate choice by nixpkgs (@sternenseemann).  The `vc4-elf` case really can't be fixed.

Reply:
```
$ /nix/store/vqs7rm7vn8v3ja3sar609zlg6xd2zczh-gnu-config-2023-01-21/config.sub x86_64-linux
x86_64-pc-linux-gnu
```

```
$ nix-instantiate --eval --expr 'with (import ./lib).systems; (elaborate { system = "x86_64-linux"; }).config'
"x86_64-unknown-linux-gnu"
```